spec/main-process/parse-command-line.test.js

```diff
@@ -21,6 +21,61 @@ describe('parseCommandLine', () => {
       ]);
       assert.deepEqual(args.pathsToOpen, ['/some/path']);
     });
+
+    describe('and --_ or -_ are passed', () => {
```

Comment:
Why single out and ignore underscore flags? I guess it's related to `-psn_...` containing an underscore? Is there a drawback to just allowing (and ignoring) it?

Reply:
_Passing `-_` or `--_`, as either the last argument, or "without a non-flag argument immediately after", is the minimal reproducer for the Gatekeeper partial startup bug._

Maybe I should add this as a code comment in the test file? Or incorporate that into the test's `describe`/`it` text?

---

`yargs` stores non-flag arguments in an array called `argv._`, but this can easily be polluted with the content of actual underscore `_` flags. (See Gatekeeper's `-psn_0_[1234567]`). And `yargs` treats any flag with "no content" associated with it (no no-flag argument passed right after the flag) as boolean `true`s.

Passing the "underscore flag" is the main way to get stray boolean `true`s into the non-flag argument pool. It's an easy and convenient way to test the Gatekeeper flag bug more generally than a flag starting with `-psn_`.

Restated in terms relative to the normal non-bug behavior: This test handles boolean pollution in the non-flag argument array, which is the one where we look for paths and URIs that Atom should open.

<details><summary>Huge long reply, since I'm not totally sure what else you may need clarifying, and I was kind of on a roll. And it's late where I am. (click to expand if there's anything else you were wondering, I may have answered it here):</summary>

_(Before I get into it: Admittedly none of the following complicated information would make sense from just looking at the test as I wrote it. So maybe I should add comments in the test file. Or make the purpose of the test more obvious from the `describe` and `it` text. TODO: make the test self-explanatory.)_

Also to address just this question directly:

> Why single out and ignore underscore flags? I guess it's related to -psn_... containing an underscore? Is there a drawback to just allowing (and ignoring) it?

I felt that it was just a matter of time before we saw a similarly-formed flag. If it didn't happen to start with exactly `-psn_`, then we wouldn't be filtering it out, and then we're back to experiencing essentially this same issue.

The general solution of handling/ignoring shortflag sequences containing effectively `-_`... is very simple. It happens to also filter out `--_`. I may have gone overboard in explicitly testing for both, but `yargs` treats them the same under the hood right now, the solution to handle all of this comprehensively is a few lines of code, and I just wanted to be thorough. I figured being mildly paranoid about it now would make our code and test coverage robust to whatever strange inputs may come, and alert us to whatever undocumented minor behavior changes may happen in `yargs`, for instance.

---

So, the reason I went there at all is a bit of a rabbit-hole of how yargs works. Well, it seems like a rabbit-hole if you don't know where it ends up. I will try to make this concise but complete.

### Background info: what's the normal, desired behavior?

Starting with the normal, desired behavior: `yargs` stores all non-flag arguments in an array called `argv._`.

So, Atom (rather reasonably) expects non-flag arguments to be filenames or URLs we want to open. For example:

Running `atom file1 ./file2 /path/to/file3` will launch Atom with tabs open for these files.

### Quirks in yargs

A few things make it complicated:
- Users and the OS can specify a literal `_` flag of either the short form `-_` or `--_`, and `yargs` will dump it right into the same `argv._` array with the "non-flag" arguments.
- `yargs` by default treats any flag with no (non-flag) argument after it as boolean `true`
- Bonus complication that's not strictly a part of the bug I'm trying to fix: Any bare number passed as a non-flag argument, say `33`, will be turned into a JS `number`. (There is a feature to treat everything in `yargs.argv._` as strings, but it didn't work for numbers in my testing. It didn't coerce numbers to strings in my testing, it left them as JS `number` types. I think this is a bug in `yargs`).

Side note: _(`yargs` seems to follow a Linux-style convention about shortflags. It's a conventional behavior on Linux to treat a string of characters after a single dash as being a number of shortflags. For example, a typical `tar` invocation: `tar -xvf my_tarball.tgz`; in English that's "tar, extract verbosely a file named `my_tarball.tgz`.")_

### The purpose of that test (filtering out booleans)

The purpose of that test is to test an array of inputs that has been polluted with booleans. This is the scenario that the `-psn_0_[1234567]` flag presents.

**_Passing `-_` or `--_`, without a non-flag argument immediately after, is the minimal reproducer for the Gatekeeper partial startup bug._**

### Why I didn't just pass booleans into the parser function

Passing non-strings to this function doesn't resemble its real-world use. The function is only used in one place. [Here in `src/main-process/start.js`](https://github.com/atom/atom/blob/v1.53.0/src/main-process/start.js#L45). Atom always, 100% of the time, passes the `parseCommandLine()` function either an empty array, or an array of strings. These are read from Node's `process.argv` immediately before passing to the `parseCommandLine()` function. Node's `process.argv` will only ever contain strings, apparently. The diverse typing come in only _after_ `yargs` has taken a look at the inputs and decided whether they are strings, numbers, booleans...

</details>

Reply:
Makes sense. I think a short comment on that test to explain why would be useful.

(turns out yargs has an open bug related to this too https://github.com/yargs/yargs/issues/878)